File: backend/src/main/java/com/calonuria/backend/dto/journal/LibroJournalRegistroDTO.java
package com.calonuria.backend.dto.journal;

import jakarta.validation.constraints.*;
import lombok.Data;
import java.time.LocalDate;
import java.util.List;

@Data
public class LibroJournalRegistroDTO {

    @NotNull(message = "El usuario es obligatorio")
    private Long idUsuario;

    // Se elimina @NotNull para permitir guardar con los datos de Google Books directamente
    private Long idLibro;

    // --- Datos del libro (Por si es la primera vez que se añade a la app) ---
    private String googleBooksId;
    private String titulo;
    private String autor;
    private String isbn;
    private String editorial;
    private String descripcion;
    private String portadaUrl;
    private String genero;
    private String tipoLibro;
    private Integer anioPublicacion;
    // ------------------------------------------------------------------------

    @NotBlank(message = "El estado es obligatorio")
    @Pattern(regexp = "Pendiente|Leyendo|Terminado|Abandonado",
            message = "Estado no válido")
    private String estado;

    @Min(value = 0, message = "La página no puede ser negativa")
    private Integer paginaActual;

    @Min(value = 1, message = "La valoración mínima es 1")
    @Max(value = 5, message = "La valoración máxima es 5")
    private Integer valoracion;

    @Pattern(regexp = "Fisico|Digital|Audiolibro",
            message = "Formato no válido")
    private String formatoLectura;

    private List<String> emociones;
    private String citasFavoritas;
    private String notaPersonal;
    private LocalDate fechaInicio;
    private LocalDate fechaFin;
}

File: frontend/between_pages/lib/models/journal/libro_journal_registro_dto.dart
class LibroJournalRegistroDTO {
  final int idUsuario;
  final int? idLibro;

  final String? googleBooksId;
  final String? titulo;
  final String? autor;
  final String? isbn;
  final String? editorial;
  final String? descripcion;
  final String? portadaUrl;
  final String? genero;
  final String? tipoLibro;
  final int? anioPublicacion;

  final String estado;
  final int? paginaActual;
  final int? valoracion;
  final String? formatoLectura;

  final List<String>? emociones;
  final String? citasFavoritas;
  final String? notaPersonal;

  final String? fechaInicio;
  final String? fechaFin;

  LibroJournalRegistroDTO({
    required this.idUsuario,
    this.idLibro,
    this.googleBooksId,
    this.titulo,
    this.autor,
    this.isbn,
    this.editorial,
    this.descripcion,
    this.portadaUrl,
    this.genero,
    this.tipoLibro,
    this.anioPublicacion,
    required this.estado,
    this.paginaActual,
    this.valoracion,
    this.formatoLectura,
    this.emociones,
    this.citasFavoritas,
    this.notaPersonal,
    this.fechaInicio,
    this.fechaFin,
  });

  factory LibroJournalRegistroDTO.fromJson(Map<String, dynamic> json) {
    return LibroJournalRegistroDTO(
      idUsuario: json['idUsuario'] as int,
      idLibro: json['idLibro'] as int?,
      googleBooksId: json['googleBooksId'] as String?,
      titulo: json['titulo'] as String?,
      autor: json['autor'] as String?,
      isbn: json['isbn'] as String?,
      editorial: json['editorial'] as String?,
      descripcion: json['descripcion'] as String?,
      portadaUrl: json['portadaUrl'] as String?,
      genero: json['genero'] as String?,
      tipoLibro: json['tipoLibro'] as String?,
      anioPublicacion: json['anioPublicacion'] as int?,
      estado: json['estado'] as String,
      paginaActual: json['paginaActual'] as int?,
      valoracion: json['valoracion'] as int?,
      formatoLectura: json['formatoLectura'] as String?,
      emociones: (json['emociones'] as List?)?.cast<String>(),
      citasFavoritas: json['citasFavoritas'] as String?,
      notaPersonal: json['notaPersonal'] as String?,
      fechaInicio: json['fechaInicio'] as String?,
      fechaFin: json['fechaFin'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'idUsuario': idUsuario,
      'idLibro': idLibro,
      'googleBooksId': googleBooksId,
      'titulo': titulo,
      'autor': autor,
      'isbn': isbn,
      'editorial': editorial,
      'descripcion': descripcion,
      'portadaUrl': portadaUrl,
      'genero': genero,
      'tipoLibro': tipoLibro,
      'anioPublicacion': anioPublicacion,
      'estado': estado,
      'paginaActual': paginaActual,
      'valoracion': valoracion,
      'formatoLectura': formatoLectura,
      'emociones': emociones,
      'citasFavoritas': citasFavoritas,
      'notaPersonal': notaPersonal,
      'fechaInicio': fechaInicio,
      'fechaFin': fechaFin,
    };
  }
}
